sdk/iot/docs/mqtt_state_machine.md

```diff
@@ -0,0 +1,223 @@
+# Azure IoT SDK v2 MQTT State Machine
+
+## High-level architecture
+
+Device Provisioning and IoT Hub service protocols require additional state management on top of the MQTT protocol. The Azure IoT SDK v2 components provide a common programming model for the C SDK.
+
+The following aspects are being handled by the SDK:
+1. Generate MQTT CONNECT credentials.
+1. Obtain SUBSCRIBE topic filters required by various service features.
+1. Output a standardized MQTT PUBLISH object.
+1. Parse service errors and output an uniform error object model (i.e. az_result) 
+1. Provide the correct sequence of events required to perform an operation. 
+1. Provide suggested timing information when retrying operations.
+
+## Components
+
+### IoT Hub
+![image](iot_hub_flow.png "IoT MQTT State Machine")
+
+### Device Provisioning Service
+![image](iot_provisioning_flow.png "Device Provisioning MQTT State Machine")
+
+### Common Retry Flow
+![image](iot_retry_flow.png "MQTT Retry Flow")
+
+## Design Decisions
+Porting requirements:
+- The target platform C99 compiler is generating reentrant code. The target platform supports a stack of several kB.
+- Our SDK relies on types such as `uint8_t` existing. If a platform doesn't already have them defined, they should be added to the application code using `typedef` statements.
+
+The SDK is provided only for the MQTT protocol. Support for WebSocket/WebProxy tunneling as well as TLS security is not handled by the SDK.
+
+The following API provides a framework and a programming model. Each service feature (telemetry, methods, etc) will have a separate client following the same structure to allow users to only consume and compile Azure IoT features that their devices require.
+
+Assumptions and recommendations for the application:
+- Our API does not support unsubscribing from any of the previously subscribed topics. We assume that the device will only set-up topics that must be used.
+- It is recommeded that MQTT layer uses the `clean_session == FALSE` during MQTT CONNECT. This allows the application to avoid resubscribing.
+
+## API
+
+### MQTT messages
+
+#### MQTT CONNECT
+```C
+typedef struct az_iot_mqtt_connect {
+    az_span client_id;  // NULL terminated string.
+    az_span user;       // NULL terminated string.
+    az_span password;   // NULL terminated string (may be an empty string).
+} az_iot_mqtt_connect;
+```
+
+#### MQTT SUB
+```C
+typedef struct az_iot_topic {
+    az_span name;       // NULL terminated string.
+    uint8_t qos;
+} az_iot_topic;
+```
+
+#### MQTT PUB
+
+`Note:` this applies for the MQTT PUB either sent or received.
+
+```C
+typedef struct az_iot_mqtt_pub {

```

Comment:
Offline discussion: proposal to remove this struct and have _get_topic instead. Payload will be managed by the application layer entirely:
- telemetry, method_response, twin_patch accept a payload
- all other request-based APIs need an empty payload